### PR TITLE
WIP: Make PDFs GPU ready / Use outer constuctors in PDFs

### PR DESCRIPTION
**Pull request by nmheim**
I am having some problems with moving the PDF structs to GPU. One of them is that `mapleaves` wants to call constructors without parametric types

```juila
using CuArrays, Flux, GenerativeModels

p = CGaussian{Float32,UnitVar}(3, 2, Dense(2,3))
p = p |> gpu
```

results in the following error
```julia
ERROR: MethodError: no method matching CGaussian(::Int64, ::Int64, ::Dense{typeof(identity),TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}}})
Stacktrace:
 [1] mapchildren(::Function, ::CGaussian{Float32,UnitVar}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:15
 [2] #mapleaves#31 at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:29 [inlined]
 [3] mapleaves at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:28 [inlined]
 [4] gpu at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:69 [inlined]
 [5] |>(::CGaussian{Float32,UnitVar}, ::typeof(gpu)) at ./operators.jl:813
 [6] top-level scope at none:0
```

redefining our constructors as in  this PR could solve that. However, then I run into another error which I have not been able to resolve yet:
```juila
using CuArrays, Flux, GenerativeModels

p = CGaussian(3, 2, Dense(2,3))
p = p |> gpu
```
results in this:
```julia
ERROR: LoadError: cannot assign variables in other modules
Stacktrace:
 [1] #9 at ./reflection.jl:684 [inlined]
 [2] ntuple(::getfield(Base, Symbol("##9#10")){DataType}, ::Int64) at ./tuple.jl:136
 [3] fieldtypes(::Type) at ./reflection.jl:684
 [4] explain_nonisbits(::Any, ::Int64) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/validation.jl:44
 [5] explain_nonisbits(::Any, ::Int64) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/validation.jl:47 (repeats 5 times)
 [6] explain_nonisbits at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/validation.jl:42 [inlined]
 [7] check_invocation(::CUDAnative.CompilerJob, ::LLVM.Function) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/validation.jl:68
 [8] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:187 [inlined]
 [9] macro expansion at /home/niklas/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
 [10] #codegen#130(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:186
 [11] #codegen at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:0 [inlined]
 [12] #compile#129(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:47
 [13] #compile#128 at ./none:0 [inlined]
 [14] #compile at ./none:0 [inlined] (repeats 2 times)
 [15] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:389 [inlined]
 [16] #cufunction#170(::Nothing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::getfield(GPUArrays, Symbol("##23#24")), ::Type{Tuple{CuArrays.CuKernelState,CUDAnative.CuDeviceArr
ay{Float32,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(convert),Tuple{CUDAnative.CuRefValue{DataType},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.A
S.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:357
 [17] cufunction(::Function, ::Type) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:357
 [18] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:174 [inlined]
 [19] macro expansion at ./gcutils.jl:87 [inlined]
 [20] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:171 [inlined]
 [21] _gpu_call(::CuArrays.CuArrayBackend, ::Function, ::CuArray{Float32,2}, ::Tuple{CuArray{Float32,2},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(convert),Tuple{Base.RefValue{Type{Float32}},Base.
Broadcast.Extruded{CuArray{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at /home/niklas/.julia/packages/CuArrays/wXQp8/src/gpuarray_interface.jl:60
 [22] gpu_call at /home/niklas/.julia/packages/GPUArrays/J4c3Q/src/abstract_gpu_interface.jl:151 [inlined]
 [23] gpu_call at /home/niklas/.julia/packages/GPUArrays/J4c3Q/src/abstract_gpu_interface.jl:128 [inlined]
 [24] copyto! at /home/niklas/.julia/packages/GPUArrays/J4c3Q/src/broadcast.jl:48 [inlined]
 [25] copyto! at ./broadcast.jl:797 [inlined]
 [26] copy at ./broadcast.jl:773 [inlined]
 [27] materialize at ./broadcast.jl:753 [inlined]
 [28] adapt_storage(::Type{Float32}, ::CuArray{Float32,2}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:73
 [29] adapt_structure(::DataType, ::CuArray{Float32,2}) at /home/niklas/.julia/packages/Adapt/aeQPS/src/Adapt.jl:9
 [30] adapt at /home/niklas/.julia/packages/Adapt/aeQPS/src/Adapt.jl:6 [inlined]
[31] adapt_structure(::DataType, ::TrackedArray{…,CuArray{Float32,2}}) at /home/niklas/.julia/packages/Tracker/SAr25/src/Tracker.jl:112
 [32] adapt at /home/niklas/.julia/packages/Adapt/aeQPS/src/Adapt.jl:6 [inlined]
 [33] #43 at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:75 [inlined]
 [34] #mapleaves#31(::IdDict{Any,Any}, ::Function, ::getfield(Flux, Symbol("##43#44")){DataType}, ::TrackedArray{…,CuArray{Float32,2}}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:29
 [35] #mapleaves at ./none:0 [inlined]
 [36] #32 at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:29 [inlined]
 [37] _broadcast_getindex_evalf(::getfield(Flux, Symbol("##32#33")){IdDict{Any,Any},getfield(Flux, Symbol("##43#44")){DataType}}, ::TrackedArray{…,CuArray{Float32,2}}) at ./broadcast.jl:578
 [38] _broadcast_getindex at ./broadcast.jl:551 [inlined]
 [39] (::getfield(Base.Broadcast, Symbol("##19#20")){Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,getfield(Flux, Symbol("##32#33")){IdDict{Any,Any},getfield(Flux, Symbol("##43#44")){DataType}},Tuple{Tuple{TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}},typeof(identity)}}}})(::Int64) at ./broadcast.jl:953
 [40] ntuple(::getfield(Base.Broadcast, Symbol("##19#20")){Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,getfield(Flux, Symbol("##32#33")){IdDict{Any,Any},getfield(Flux, Symbol("##43#44")){DataType}},Tuple{Tuple{TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}},typeof(identity)}}}}, ::Val{3}) at ./tuple.jl:161
 [41] copy at ./broadcast.jl:953 [inlined]
 [42] materialize at ./broadcast.jl:753 [inlined]
 [43] mapchildren(::Function, ::Dense{typeof(identity),TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}}}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:15
 [44] #mapleaves#31(::IdDict{Any,Any}, ::Function, ::getfield(Flux, Symbol("##43#44")){DataType}, ::Dense{typeof(identity),TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}}}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:29
 [45] mapleaves at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:28 [inlined]
 [46] paramtype(::Type{Float32}, ::Dense{typeof(identity),TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}}}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:75
 [47] CGaussian(::Int64, ::Int64, ::Dense{typeof(identity),TrackedArray{…,CuArray{Float32,2}},TrackedArray{…,CuArray{Float32,1}}}, ::Type) at /home/niklas/repos/GenerativeModels.jl/src/pdfs/cgaussian.jl:74 (repeats 2 times)
 [48] mapchildren(::Function, ::CGaussian{Float32,UnitVar}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:15
 [49] #mapleaves#31(::IdDict{Any,Any}, ::Function, ::typeof(cu), ::CGaussian{Float32,UnitVar}) at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:29
 [50] mapleaves at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:28 [inlined]
 [51] gpu at /home/niklas/.julia/packages/Flux/dkJUV/src/treelike.jl:69 [inlined]
 [52] |>(::CGaussian{Float32,UnitVar}, ::typeof(gpu)) at ./operators.jl:813
 [53] top-level scope at none:0
 [54] include at ./boot.jl:326 [inlined]
 [55] include_relative(::Module, ::String) at ./loading.jl:1038
 [56] include(::Module, ::String) at ./sysimg.jl:29
 [57] include(::String) at ./client.jl:403
in expression starting at /home/niklas/repos/Rodent/scripts/cuda_test.jl:32
```

### CODE REVIEW

**Comment by vitskvara** [comment]
Can you maybe only move the mappings to gpu?

Dne čt 19. 9. 2019 15:21 uživatel Niklas Heim <notifications@github.com>
napsal:

**Comment by vitskvara** [comment]
Or you could overload the gpu macro for cpdfs soit onyl affects the mapping.

Dne čt 19. 9. 2019 15:51 uživatel Vít Škvára <vit.skvara@gmail.com> napsal:

**Comment by nmheim** [comment]
ok, so the problem was that I tried to convert the mapping, which aparently CuArrays don't like:
https://github.com/nmheim/GenerativeModels.jl/blob/426efc6790ce73b7131982c1860a67c92de89aef/src/pdfs/cgaussian.jl#L51-L57

I removed the conversion and instead the parametric type of the `CGaussian` is not inferred in the constructor.
```julia
T = eltype(first(params(mapping)).data)
V = _detect_mapping_variant(mapping, xlength, zlength)
```
This also makes it easier to create cpdfs:
```julia
p = CGaussian(3, 2, Dense(2, 3))
# instead of this:
p = CGaussian{Float64,UnitVar}(3, 2, Dense(2, 3))
```
The old, longer constuctors are still made available by Julia but are not doing any checks anymore. I assume we are fine with this? @vitskvara

**Comment by vitskvara** [comment]
Yeah, I am fine with this.

**Comment by nmheim** [comment]
Runnig into problems with the `likelihood` function of the `CGaussian`:
```julia
using CuArrays
using Flux
using GenerativeModels

p = CGaussian(2, 2, x->tanh.(x))
q = CGaussian(2, 2, x->tanh.(x))
vae = VAE(p,q) |> gpu

z  = randn(Float32, 2, 100) |> gpu
x  = randn(Float32, 2, 100) |> gpu
σ2 = variance(p, z)
μx = mean(p, z)
μx = mean(q, z)

llh = mean(-loglikelihood(vae.decoder, x, z))
```
Produces:
```julia
ERROR: GPU compilation of #23(CuArrays.CuKernelState, CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(/),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{CUDAnative.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}},CUDAnative.CuRefValue{Val{2}}}},Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}}}}) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(/),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{CUDAnative.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}},CUDAnative.CuRefValue{Val{2}}}},Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}}}}.
That type is not isbits, and such arguments are only allowed when they are unused by the kernel.  .args is of type Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{CUDAnative.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}},CUDAnative.CuRefValue{Val{2}}}},Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}}} which is not isbits.
    .2 is of type Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}} which is not isbits.
      .x is of type Array{Float32,1} which is not isbits.


Stacktrace:
 [1] check_invocation(::CUDAnative.CompilerJob, ::LLVM.Function) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/validation.jl:70
 [2] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:187 [inlined]
 [3] macro expansion at /home/niklas/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
 [4] #codegen#130(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:186
 [5] #codegen at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:0 [inlined]
 [6] #compile#129(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/compiler/driver.jl:47
 [7] #compile#128 at ./none:0 [inlined]
 [8] #compile at ./none:0 [inlined] (repeats 2 times)
 [9] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:389 [inlined]
 [10] #cufunction#170(::Nothing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::getfield(GPUArrays, Symbol("##23#24")), ::Type{Tuple{CuArrays.CuKernelState,CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(/),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{CUDAnative.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}},CUDAnative.CuRefValue{Val{2}}}},Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}}}}}}) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:357
 [11] cufunction(::Function, ::Type) at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:357
 [12] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:174 [inlined]
 [13] macro expansion at ./gcutils.jl:87 [inlined]
 [14] macro expansion at /home/niklas/.julia/packages/CUDAnative/UWBIY/src/execution.jl:171 [inlined]
 [15] _gpu_call(::CuArrays.CuArrayBackend, ::Function, ::CuArray{Float32,2}, ::Tuple{CuArray{Float32,2},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(/),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{Base.RefValue{typeof(^)},Base.Broadcast.Extruded{CuArray{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}},Base.RefValue{Val{2}}}},Base.Broadcast.Extruded{Array{Float32,1},Tuple{Bool},Tuple{Int64}}}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at /home/niklas/.julia/packages/CuArrays/wXQp8/src/gpuarray_interface.jl:60
 [16] gpu_call at /home/niklas/.julia/packages/GPUArrays/fLiQ1/src/abstract_gpu_interface.jl:151 [inlined]
 [17] gpu_call at /home/niklas/.julia/packages/GPUArrays/fLiQ1/src/abstract_gpu_interface.jl:128 [inlined]
 [18] copyto! at /home/niklas/.julia/packages/GPUArrays/fLiQ1/src/broadcast.jl:48 [inlined]
 [19] copyto! at ./broadcast.jl:797 [inlined]
 [20] copy at ./broadcast.jl:773 [inlined]
 [21] materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(/),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{CuArray},Nothing,typeof(CuArrays.culiteral_pow),Tuple{Base.RefValue{typeof(^)},CuArray{Float32,2},Base.RefValue{Val{2}}}},Array{Float32,1}}}) at ./broadcast.jl:753
 [22] loglikelihood(::CGaussian{Float32,UnitVar}, ::CuArray{Float32,2}, ::CuArray{Float32,2}) at /home/niklas/repos/GenerativeModels.jl/src/pdfs/abstract_cgaussian.jl:16
 [23] top-level scope at none:0
```
@vitskvara is this something you have already fixed in your likelihoods?

**Comment by nmheim** [comment]
nevermind, I found the error, it was amongst others related to `CGaussian` with unit variance not returning a `CuArray`. is there a better solution than this?
```julia
function mean_var(p::CGaussian{T,UnitVar}, z::AbstractArray) where T
    μ = p.mapping(z)
    σ2 = ones(T, xlength(p))
    σ2 = isa(μ, Array) ? σ2 : σ2 |> gpu
    return μ, σ2
end
```
this would have to be done in a couple of places in the code, e.g. in `rand` where we introduce a new random array...

File: src/pdfs/abstract_cgaussian.jl [hunk 1]
```diff
@@ -13,7 +13,8 @@ end
 
 function loglikelihood(p::AbstractCGaussian{T}, x::AbstractArray, z::AbstractArray) where T
     (μ, σ2) = mean_var(p, z)
-    y = (x - μ).^2
+    d = x - μ
+    y = d .* d
```
Comment (nmheim) [inline src/pdfs/abstract_cgaussian.jl:17]: Is this really an issue for everyone or can this be related to some old CUDA version or so?
`.^2` complains for me that scalar getindex is disallowed

**Pull request merged**